### PR TITLE
Concurrent Consumer creation

### PR DESCRIPTION
**Pull request by jmillan**
Create Consumers concurrently.

When there is an ongoing Consumer creation, store the subsequent Consumer creation tasks and run them at once after the current one is created.

This allows operating RTCPeerConnection a single time for multiple track reception rather than doing it sequentially one by one, with the corresponding speed gain.

### CODE REVIEW

**Comment by msach22** [comment]
🔥 🔥

File: src/Transport.ts [hunk 1]
```diff
@@ -19,6 +19,24 @@ interface InternalTransportOptions extends TransportOptions
 	canProduceByKind: CanProduceByKind;
 }
 
+class ConsumerCreationTask
+{
+	consumerOptions: ConsumerOptions;
+	promise: Promise<Consumer>;
+	resolve?: (consumer: Consumer) => void;
+	reject?: (error: Error) => void;
```
Comment (ibc) [inline src/Transport.ts:27]: Shouldn't some of those be `private`?

File: src/Transport.ts [hunk 2]
```diff
@@ -788,6 +775,124 @@ export class Transport extends EnhancedEventEmitter
...
+					const results = await this._handler.receive(optionsList);
+
+					for (let idx=0; idx < results.length; idx++)
+					{
+						const task = pendingConsumerTasks[idx];
+						const result = results[idx];
+						const { id, producerId, kind, rtpParameters, appData } = task.consumerOptions;
+
+						const { localId, rtpReceiver, track } = result;
+						const consumer = new Consumer(
```
Comment (ibc) [inline src/Transport.ts]:
```suggestion
						const task = pendingConsumerTasks[idx];
						const result = results[idx];
						const { id, producerId, kind, rtpParameters, appData } = task.consumerOptions;
						const { localId, rtpReceiver, track } = result;
						const consumer = new Consumer(
```

File: src/Transport.ts [hunk 3]
```diff
@@ -788,6 +775,124 @@ export class Transport extends EnhancedEventEmitter
 			'transport.consumeData()');
 	}
 
+	// This method is guaranteed to never throw.
+	async _createPendingConsumers(): Promise<void>
+	{
+		this._awaitQueue.push(
+			async () =>
+			{
+				this._consumerCreationInProgress = true;
```
Comment (ibc) [inline src/Transport.ts:784]: Do we really need this flag? Couldn't we just rely on `this._pendingConsumerTasks.length`?

File: src/Transport.ts (hunk 1 above)
Comment (jmillan) [reply src/Transport.ts:27]: nop, we need to access all of them from `Transport`

File: src/Transport.ts (hunk 3 above)
Comment (jmillan) [reply src/Transport.ts:784]: Imagine we are creating the 1st `Consumer`.  Before the `Consumer` creation `Transport.consume()` get's called again. `pendingConsumerTasks.length` would be 0 and we would, thus, not add this task to the array but call `handler.receive` as before.

**Review by ibc — APPROVED** [review]

**Pull request merged**